manifest/main.js

```diff
@@ -0,0 +1,127 @@
+const gzip = promisify(zlib.gzip)
+
+/**
+ * We define a number of data sources, each of which represent an s3 inventory
+ * or GitHub repo listing (not yet implemented). Each of these data sources
+ * provides functions to group items (files) into resources, each with an
+ * resourceType & id. This decouples the data sources here from the sources used
+ * in the main nextstrain.org codebase.
+ */
+const DATA_SOURCES = [
+  coreData,
+  stagingData,
+];
+
+function parseArgs() {
+  const argparser = new ArgumentParser({
+    description: `
+      Fetch file lists from a number of data sources (such as S3 inventories, GitHub repo contents) and
+      collect them into resources. The collection of resources is hierarchical, grouped first by resource type
+      ('dataset', 'narrative', 'intermediate') and then by the ID of the resource. For dataset and narrative
+      resources the ID corresponds to the nextstrain.org path for the latest available version.
+      For each resource ID, we list URLs for the current version and all previous versions, where available.
+      The output JSON is intended for consumption by the nextstrain.org server.
+    `,
+  });
+  argparser.addArgument("--input",  {metavar: "<json>", type: "string", defaultValue: "",
+    help: "Optional input JSON (i.e. the output of a previous run). This is useful for incremental updates"})
+  argparser.addArgument("--local", {action: 'storeTrue',
+    help: 'Access a local copy of S3 inventories within ./devData/. See docstring of fetchInventoryLocal() for expected filenames.'})
+  argparser.addArgument("--restrict-sources", {metavar: "<name>", type: "string", nargs: '+',
+    help: "Only fetch data from a subset of sources. Source names are those defined in DATA_SOURCES"});
+  argparser.addArgument("--output", {metavar: "<json>", required: true})
+  argparser.addArgument("--indent", {action: 'storeTrue', help: 'Indent the output JSON'})
+  argparser.addArgument("--gzip", {action: 'storeTrue', help: 'GZip the output JSON'})
+
+  return argparser.parseArgs();
+}
+
+
+main(parseArgs());
+
+
+async function main(args) {
+
+  // load existing resources from disk, if requested (useful if we update one source at a time)
+  const resources = args.input ?
+    JSON.parse(fs.readFileSync(args.input, {encoding: 'utf8', flag: 'r'})) :
+    {};
+
+  for (const src of DATA_SOURCES) {
+    if (args.restrict_sources && !args.restrict_sources.includes(src.name)) {
+      continue
+    }
+
+    if (src.type !== 's3Inventory') {
+      logger.error(`Data source provided for a type we don't yet handle (${src.type})`)
+    }
+
+    /* Before we fetch the (updated) data and add it into the overall collection of resources,
+    delete any existing data from this data source */
+    for (const resourceType of Object.keys(resources)) {
+      for (const [id, resource] of Object.entries(resources[resourceType])) {
+        if (src.removeResource(resourceType, id, resource)) {
```

Comment:
Why dispatch this condition to the per-source config, esp. when both existing source configs just compare to source name? It seems convoluted for no reason.

Reply:
My original desire was to allow incremental index creation on a data-source-by-data-source basis, which would require some way of removing previous entries. In such a design I think it's cleaner to delegate this functionality to the data source itself, even if it is a one-liner.

Given that this PR only implements two data-sources I think it's better to leave the incremental update functionality to a later PR, if and when it's desired.

Reply:
I would think to do this incremental creation not by merging existing contents with new contents, which is fraught with issues (deleting, locking, and more), but by splitting the index into separate files. Splitting by source, for example, and then storing the indexes within each source. This would also be a natural way to address the keying/partitioning by source I suggested elsewhere and in our call yesterday.

Reply:
For the current (and foreseeable future) I think having a single monolithic index file is simpler to reason with, and simpler to maintain. The structure of this index has been changed to be keyed by source. If in the future we find it necessary to switch to separate index files then that change shouldn't be problematic to implement.

Reply:
Alright. I expect we're going to want that partitioning in the future, but yes, it can wait.